citizennet/purescript-ocelot
# FBCM-4151 Allow disabling Dropdowns dynamically

## What does this pull request do?

We need to be able to toggle a Dropdown's disabled status after it's been initialized. Disabling Dropdowns was never explicitly supported, and only happened to sort of work thanks to the browser disabling behavior for elements with the `disabled` attribute. We promote disabled status to a first-class citizen for Dropdowns, and do some refactoring in the process.

## Where should the reviewer start?

Commit-by-commit.

## How should this be manually tested?

1. Run `make build-ui` and navigate to the Dropdowns page of the UI Guide (if testing locally, open `dist/index.html` in your browser, otherwise see `.k8s/README.md` for instructions on deploying the UI Guide on EC2).
2. [ ] Toggling the button next to any of the disabled Dropdowns should toggle their disabled status.

## Other Notes:

The changes here constitute a breaking change. If you want to validate these changes in Wildcat or OMS, you can:

1. In the project's `packages.dhall`, change the Ocelot version to:
    ```diff
    - , version = "v0.26.2"
    + , version = "origin/FBCM-4151/davezuch/allow-disabling-dropdowns-dynamically"
    ```
1. In Wildcat, run these `sed` commands to change the relevant module names:
    ```sh
    find web/purs/src -type f -name '*.purs' | xargs sed -i 's/Ocelot.Components.Dropdown.Component/Ocelot.Dropdown/g'
    find web/purs/src -type f -name '*.purs' | xargs sed -i 's/Ocelot.Components.Dropdown.Render/Ocelot.Dropdown/g'
    ```
   In OMS, the `sed` commands should be:
    ```sh
    find infrastructure/src -type f -name '*.purs' | xargs sed -i 's/Ocelot.Components.Dropdown.Component/Ocelot.Dropdown/g'
    find infrastructure/src -type f -name '*.purs' | xargs sed -i 's/Ocelot.Components.Dropdown.Render/Ocelot.Dropdown/g'
    ```
1. You'll have to remove duplicate imports in a few modules since both Dropdown modules have now been combined into one.
1. Finally, now that we've added `disabled` to the Dropdown's input, you'll need to update the input for all the components in the project.

After the above steps, the project should build successfully.
